src/parse/lex.rs

```diff
@@ -1,22 +1,319 @@
-use super::token::{Token, TokenSpan};
+use super::token::{Token, TokenSpan, Lit};
+use std::str::Chars;
 use super::Span;
+use std::iter::{Iterator};
 
+///A lexer with its associated query, a char iterator, and
+///positions (last, current, next)
 pub struct Lexer<'a> {
     query: &'a str,
+    chs : Chars<'a>,
+    last: Option<char>,
+    last_pos: Option<usize>,
     curr: Option<char>,
-    curr_pos: Option<u32>,
+    curr_pos: Option<usize>,
+    next: Option<char>,
+    span_start: Option<usize>
 }
 
 impl<'a> Lexer<'a> {
+
+    ///new lexer, everything set to None
     pub fn from_query<'b>(query: &'b str) -> Lexer<'b> {
-        Lexer {
+        let mut lex = Lexer {
             query: query,
             curr: None,
-            curr_pos: None
+            curr_pos: Some(0),
+            last: None,
+            last_pos: Some(0),
+            next: None,
+            span_start: Some(0),
```

Comment:
I don't understand why the default value for the `pos` fields is not `None`, but `Some(0)`...